common/src/Microsoft.Azure.IIoT.Cdm/src/Services/CdmMessageProcessor.cs

```diff
@@ -116,49 +127,101 @@ public class CdmMessageProcessor : ICdmClient {
+                if (Manifest == null) {
+                    _logger.Warning("Manifest is not assigned yet. Retry ... ");
+                    await OpenAsync();
+                }
+                var entityDeclaration = Manifest.Entities.Item(kPublisherSampleEntityName);
+                if (entityDeclaration == null) {
+                    // failed to load the cdm model
+                    _logger.Error("Failed to load the entity declaration");
+                }
+                else {
+                    var partition = entityDeclaration.DataPartitions[0];
+                    var csvTrait = partition.ExhibitsTraits.Item("is.partition.format.CSV");
+                    var partitionLocation = _cdmCorpus.Storage.CorpusPathToAdapterPath(partition.Location);
+
+                    await _storage.WriteInCsvPartition<SubscriberCdmSampleModel>(
+                        partitionLocation,
+                        _cacheList,
+                        csvTrait?.Arguments?.FetchValue("delimiter") ?? kCsvPartitionsDelimiter);
+                }
+                _logger.Information("Successfully sent processed CDM data {count} records (took {elapsed}).", 
+                    _cacheList.Count, sw.Elapsed);
+                _cacheList.Clear();
             }
-
-            var entityDeclaration = Manifest.Entities.Item(kPublisherSampleEntityName);
-            if (entityDeclaration == null) {
-                // failed to load the cdm model
-                _logger.Error("Failed to load the entity declaration");
-                return;
+            catch (Exception e) {
+                var errorMessage = e.Message;
+                if (e.InnerException != null) {
+                    errorMessage += " - " + e.InnerException.Message;
+                }
+                _logger.Warning("Failed to send processed CDM data after {elapsed} : {message}",
+                     sw.Elapsed, errorMessage);
+            }
+            finally {
+                _lock.Release();
             }
+        }
 
-            var partition = entityDeclaration.DataPartitions[0];
-            var csvTrait = partition.ExhibitsTraits.Item("is.partition.format.CSV");
-            var partitionLocation = _cdmCorpus.Storage.CorpusPathToAdapterPath(partition.Location);
+        /// <summary>
+        /// Cache Timer Elapesed handler 
+        /// </summary>
+        /// <param name="sender"></param>
+        private async void CacheTimer_ElapesedAsync(object sender) {
+            await PerformWriteCache();
+            Try.Op(() => _cacheUploadTimer.Change(_cacheUploadInterval, Timeout.InfiniteTimeSpan));
+        }
 
-            await _storage.WriteInCsvPartition<SubscriberCdmSampleModel>(
-                partitionLocation,
-                new List<SubscriberCdmSampleModel>() { payload },
-                csvTrait?.Arguments?.FetchValue("delimiter") ?? kCsvPartitionsDelimiter);
+        private async Task ProcessCdmSampleAsync(SubscriberCdmSampleModel payload) {
```

Comment:
Awesome - and ok for now - however - we must make this a full blown IDeviceTelemetryHandler, that way we can do the checkpointing after the batch write.   The current way we would loose the content of the cache if we go down causing data loss.